hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieCompactedLogRecordScanner.java

```diff
@@ -63,69 +68,96 @@
     Map<String, HoodieRecord<HoodieAvroPayload>> records = Maps.newHashMap();
     // iterate over the paths
-    logFilePaths.stream().map(s -> new HoodieLogFile(new Path(s))).forEach(s -> {
-      log.info("Scanning log file " + s.getPath());
+    Iterator<String> logFilePathsItr = logFilePaths.iterator();
+    while(logFilePathsItr.hasNext()) {
+      HoodieLogFile logFile = new HoodieLogFile(new Path(logFilePathsItr.next()));
+      log.info("Scanning log file " + logFile.getPath());
       totalLogFiles.incrementAndGet();
       try {
         // Use the HoodieLogFormatReader to iterate through the blocks in the log file
-        HoodieLogFormatReader reader = new HoodieLogFormatReader(fs, s, readerSchema);
-        // Store the records loaded from the last data block (needed to implement rollback)
-        Map<String, HoodieRecord<HoodieAvroPayload>> recordsFromLastBlock = Maps.newHashMap();
-        reader.forEachRemaining(r -> {
+        HoodieLogFormatReader reader = new HoodieLogFormatReader(fs, logFile, readerSchema, true);
+        while(reader.hasNext()) {
+          HoodieLogBlock r = reader.next();
+          String blockInstantTime = r.getMetadata().get(HoodieMetadataType.INSTANT_TIME);
+          if(!HoodieTimeline.compareTimestamps(blockInstantTime, this.latestInstantTime,
+                  HoodieTimeline.LESSER_OR_EQUAL)) {
+            //hit a block with instant time greater than should be processed, stop processing further
+            break;
+          }
           switch (r.getBlockType()) {
             case AVRO_DATA_BLOCK:
-              log.info("Reading a data block from file " + s.getPath());
+              log.info("Reading a data block from file " + logFile.getPath());
               // If this is a avro data block, then merge the last block records into the main result
-              merge(records, recordsFromLastBlock);
-              // Load the merged records into recordsFromLastBlock
-              HoodieAvroDataBlock dataBlock = (HoodieAvroDataBlock) r;
-              loadRecordsFromBlock(dataBlock, recordsFromLastBlock);
+              merge(records, lastBlocks);
+              // store the last block
+              lastBlocks.push(r);
               break;
             case DELETE_BLOCK:
-              log.info("Reading a delete block from file " + s.getPath());
-              // This is a delete block, so lets merge any records from previous data block
-              merge(records, recordsFromLastBlock);
-              // Delete the keys listed as to be deleted
-              HoodieDeleteBlock deleteBlock = (HoodieDeleteBlock) r;
-              Arrays.stream(deleteBlock.getKeysToDelete()).forEach(records::remove);
+              log.info("Reading a delete block from file " + logFile.getPath());
+              String lastBlockInstantTime = lastBlocks.peek().getMetadata().get(HoodieMetadataType.INSTANT_TIME);
+              if(lastBlockInstantTime.equals(blockInstantTime)) {
```

Comment:
please switch the if condition and re work the comments accordingly, to avoid empty if block

Reply:
addressed.